teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: run
+spec:
+  pipelineRef:
+    name: build-test-deply
+  workspaces:
+  - name: source-code
+    persistentVolumeClaim:
+      claimName: otherpvc
+  - name: test-results
+    persistentVolumeClaim:
+      claimName: mypvc
+```
+
+Pros:
+* Allows runtime decisions about scheduling without changing the Pod
+
+Cons:
+* If it's important for a Pipeline to be executed in a certain way, that information will have to be encoded somewhere
+  other than the Pipeline 
+
+#### PipelineRun: flag
+
+This is similar to the `emptyDir` based solution but instead of adding extra meaning to `emptyDir` we add a flag to the
+runtime workspace information or to the entire PipelineRun (maybe when this flag is set workspaces do not need to be
+provided.)
+
+#### Controller level
+
+This option is [TEP-0046](https://github.com/tektoncd/community/pull/318). In this option, the Tekton controller can
+be configured to always execute Pipelines inside one pod.
+
+Pros:
+* Making it possible to execute a Pipeline in a pod will also pave the way to be able to support use cases such as
+  [local execution](https://github.com/tektoncd/pipeline/issues/235)
+* Users don't have to think about how the Pipeline will eb executed
+* Pipelines can be used without updates
+
+Cons:
+* Only cluster administrators will be able to control this scheduling, there will be no runtime or authoring time
+  flexibility
+* Executing a pipeline in a pod will require significantly re-architecting our graph logic so it can execute outside
+  the controller and has a lot of gotchas we'll need to iron out (see
+  [https://hackmd.io/@vdemeester/SkPFtAQXd](https://hackmd.io/@vdemeester/SkPFtAQXd) for some more brainstorming)
+
+### Within the Task
+
+In this option we ignore the "not at Task authoring time" requirement and we allow for Tasks to contain other Tasks.
+
+This is similar to [TEP-0054](https://github.com/tektoncd/community/pull/369) which proposes this via the Task spec
+in a Pipeline, but does not (yet) propose it for Tasks outside of Pipelines.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: build-test-upload
+spec:
+  workspaces:
+  - name: source
+    mountPath: /workspace/source/go/src/github.com/GoogleContainerTools/skaffold
+  steps:
+  - name: get-source
+    uses: git-clone
+    params:
+      url: $(params.url)
+    workspaces:
+    - name: source
+      workspace: source
+  - name: run-tests
+    image: golang
+    workingDir: $(workspaces.source.path)
+    script: |
+      go test <stuff here>
+  - name: upload-results
+    uses: gcs-upload
+```
+
+Pros:
+* Doesn't require many new concepts
+
+Cons:
+* Can create confusing chains of nested Tasks (Task A can contain Task B which can Contain Task C...)
+* Requires creating new Tasks to leverage the reuse (maybe embedded specs negate this?)
+* Doesn't help with parallel use cases
```

Comment:
Related to TEP-0046 right ? I feel most of the approach running multiple task in a Pod, will come with either a big refactoring of what the entrypoint does (if we keep parallel support) *or* a trade-off, which is "no parallel execution".

Reply:
Parallel Steps was the very first idea to address the parallelism-problem initially in may 2020. That would also require entrypoint work.  https://github.com/tektoncd/pipeline/issues/2586